docs/fsharp-cheatsheet.md

```diff
@@ -540,17 +538,156 @@ Another way of implementing interfaces is to use *object expressions*.
+| Record                                                            | `type internal MyRecord = { id: int }`     |
+| Record [ctor](#CodeOrganization_PrivateConstructors)              | `type MyRecord = private { id: int }`      |
+| Discriminated Union                                               | `type private MyDiscUni = A \| B`          |
+| Discriminated Union [ctor](#CodeOrganization_PrivateConstructors) | `type MyDiscUni = internal A \| B `        |
+| Class                                                             | `type internal MyClass() =`                |
+| Class [ctor](#CodeOrganization_PrivateConstructors)               | `type MyClass private () =`                |
+| Class Additional [ctor](#CodeOrganization_PrivateConstructors)    | `internal new() = MyClass("defaultValue")` |
+| Class .. `let`                                                    | *Always private. Cannot be modified*       |
+| `type` .. `member`                                                | `member private _.classMember =`           |
+
+<a name="CodeOrganization_PrivateConstructors"></a>
+##### Private Constructors
+
+Limiting `type` constructors (ctor) accessibility is a good way to enforce value integrity.
+
+Example of Single-case Discriminated Union with a `private` constructor:
+
+    type UnitQuantity =
+        private UnitQuantity of int
+        with
+        static member private MaxQty = 100
+        static member Create (qty:int) : Result<UnitQuantity, string> =
+            if qty <= UnitQuantity.MaxQty
+            then Ok (UnitQuantity qty)
+            else Error $"UnitQuantity cannot be more than {UnitQuantity.MaxQty}"
+    
+    let uQty = UnitQuantity.Create 50
+    match uQty with
+    | Ok (UnitQuantity qty) -> printfn $"Good: {qty}"
+    | Error errStr -> printfn $"Bad: {errStr}"
+
+Example of a class with a `private` constructor:
+
+    type MyClass private (count:int) =
+        member this.Count = count
+        static member CreateInstance (cnt: int) : MyClass option =
+            if cnt > 0
+            then Some(new MyClass(cnt))
+            else None
+
+    let myClass = MyClass.CreateInstance (5)
+
+### Recursive Reference
+
+F#'s type inference and name resolution runs in file and line order; by default, any forward references are considered as errors.
+This default provides a single benefit, which can be hard to appreciate initially: you never need to look beyond the current file for a dependency.
+In general this also nudges toward more careful design and organisation of codebases,
+which results in cleaner, maintainable code. However, in rare cases you may need to loosen those rules.
+To do this we have `rec` for `module` and `namespace`s; and `and` for `type` and [`let`](#Functions_RecursiveFunctions) functions.
+
+    module rec MyNamespace.MonkeyDomain
+
+    exception DoNotSqueezeBananaException of Banana // `Banana` has not been defined yet, and would fail without `rec`
```

Comment:
this is a valid example, but very esoteric

I'd say most common usage of `rec` is people saying "well for this file, I just want to list everything alphabetically, regardless of the fact that it's normally good to put low level stuff first and then have the hihger level stuff at the bottom, with some exceptions where you use `and` to put some irrelevant dependencies underneath"

So while this does illustrate what it does and how to use it, I dont think it conveys what it's for really, so am pretty conflicted.

Reply:
I stole this from [MS Learn](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/modules#recursive-modules). I kind of hate it too. The long names get in the way of what's being demonstrated as well.

Reply:
I took a walk and realized I'm conflicted because I hate exceptions. :)

Reply:
whereas I think it should have a link to https://eiriktsarpalis.wordpress.com/2017/02/19/youre-better-off-using-exceptions/ and https://fsharpforfunandprofit.com/posts/against-railway-oriented-programming/ :P